### PR TITLE
Install with pip

### PR DESCRIPTION
**Pull request by applenob**
Add pip install for package delta.

Considering the complexity of speech code dependencies, we extract `delta-nlp`
for pip install now. 

Users can still use the install bash script to install from our source code with both speech and nlp supported.

### CODE REVIEW

File: delta/data/preprocess/text_ops.py [hunk 1]
```diff
@@ -54,43 +53,17 @@ def tokenize_sentence(texts, max_seq_len, vocab_path):
 
 def chinese_word_cut_tf(input_str, use_file=False):
   """"""
```
Comment (zh794390558) [inline delta/data/preprocess/text_ops.py]: make sure this ok.

File: delta/layers/ops/kernels/jieba_op_test.py [hunk 2]
```diff
@@ -43,54 +43,19 @@ class JiebaOpsTest(tf.test.TestCase):
   #pylint: disable=no-self-use
   def build_op_use_file(self, sentence):
     ''' build graph '''
```
Comment (zh794390558) [inline delta/layers/ops/kernels/jieba_op_test.py]: make sure this ok.

File: delta/layers/ops/py_x_ops.py [hunk 3]
```diff
@@ -15,14 +15,23 @@
 # ==============================================================================
 ''' python custom ops '''
 
+import os
 import delta.compat as tf
 from absl import logging
 
+from delta import PACKAGE_ROOT_DIR
+from delta.data.utils import read_lines_from_text_file
```
Comment (zh794390558) [inline delta/layers/ops/py_x_ops.py]: remove this line.

File: delta/layers/ops/py_x_ops.py [hunk 4]
```diff
@@ -40,6 +49,50 @@
 vocab_id_to_token = gen_x_ops.vocab_id_to_token
 token_in_vocab = gen_x_ops.token_in_vocab
 str_lower = gen_x_ops.str_lower
-jieba_cut = gen_x_ops.jieba_cut
 sentence_to_ids = gen_x_ops.sentence_to_ids
 delta_delta = gen_x_ops.delta_delta
+
+
+def jieba_cut(input_sentence,
```
Comment (zh794390558) [inline delta/layers/ops/py_x_ops.py]: move jieba_cut out of the file.

File: delta/resources/cppjieba_dict [hunk 5]
```diff
@@ -0,0 +1 @@
+../../tools/cppjieba/dict
```
Comment (zh794390558) [inline delta/resources/cppjieba_dict]: what is this ?

File: delta/utils/register.py [hunk 6]
```diff
@@ -134,14 +134,14 @@ def __init__(self):
...
 
-ALL_MODULES = [("delta.data.task", TASK_MODULES),
-               ("delta.models", MODLE_MODULES),
-               ("delta.utils.loss", LOSS_MODULES),
-               ("delta.utils.metrics", METRICS_MODULES),
-               ("delta.utils.solver", SOLVER_MODULES),
-               ("delta.utils.postprocess", POSTPROCESS_MODULES),
-               ("delta.serving", SERVING_MODULES),
-               ("delta.data.preprocess", PREPROCESS_MODULES)]
+ALL_SYS_MODULES = [("delta.data.task", TASK_MODULES),
```
Comment (zh794390558) [inline delta/utils/register.py]: Maybe I will change this!

File: setup.py [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+import sys
+from glob import glob
+import tensorflow as tf
+
+TF_INCLUDE, TF_CFLAG = tf.sysconfig.get_compile_flags()
+TF_INCLUDE = TF_INCLUDE.split('-I')[1]
+
+TF_LIB_INC, TF_SO_LIB = tf.sysconfig.get_link_flags()
+TF_SO_LIB = TF_SO_LIB.replace('-l:libtensorflow_framework.1.dylib',
+                              '-ltensorflow_framework.1')
```
Comment (zh794390558) [inline setup.py]: tensorflow_framework.2

File: setup.py [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+TF_LIB_INC = TF_LIB_INC.split('-L')[1]
+TF_SO_LIB = TF_SO_LIB.split('-l')[1]
+
+print("TF_INCLUDE: {}".format(TF_INCLUDE))
+print("TF_CFLAG: {}".format(TF_CFLAG))
+print("TF_LIB_INC: {}".format(TF_LIB_INC))
+print("TF_SO_LIB: {}".format(TF_SO_LIB))
+
+NAME = "delta-didi"
```
Comment (zh794390558) [inline setup.py]: delta

File: setup.py [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+URL = "https://github.com/{0}/{1}".format(GITHUB_USER_NAME, REPO_NAME)
+GITHUB_RELEASE_TAG = str(date.today())
+DOWNLOAD_URL = "https://github.com/{0}/{1}/tarball/{2}".format(
+    GITHUB_USER_NAME, REPO_NAME, GITHUB_RELEASE_TAG)
+SHORT_DESCRIPTION = "DELTA is a deep learning based natural language and speech processing platform."
+PLATFORMS = ["Windows", "MacOS", "Unix"]
+CLASSIFIERS = [
+    "Development Status :: 4 - Beta",
+    "Intended Audience :: Education",
+    "License :: OSI Approved :: MIT License",
```
Comment (zh794390558) [inline setup.py]: Apache

File: setup.py [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+    "Natural Language :: English",
+    "Operating System :: MacOS",
+    "Operating System :: POSIX :: Linux",
+    "Programming Language :: Python :: 3.6",
```
Comment (zh794390558) [inline setup.py]: maybe changed.

File: setup.py [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+REPO_NAME = os.path.basename(os.getcwd())
+URL = "https://github.com/{0}/{1}".format(GITHUB_USER_NAME, REPO_NAME)
+GITHUB_RELEASE_TAG = str(date.today())
+DOWNLOAD_URL = "https://github.com/{0}/{1}/tarball/{2}".format(
+    GITHUB_USER_NAME, REPO_NAME, GITHUB_RELEASE_TAG)
+SHORT_DESCRIPTION = "DELTA is a deep learning based natural language and speech processing platform."
+PLATFORMS = ["Windows", "MacOS", "Unix"]
+CLASSIFIERS = [
+    "Development Status :: 4 - Beta",
+    "Intended Audience :: Education",
```
Comment (zh794390558) [inline setup.py]: Education?

File: setup.py [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+  except:
+    print("'requirements.txt' not found!")
+    requires = list()
+  return requires
+
+
+complie_args = [TF_CFLAG, "-fPIC", "-shared", "-O2", "-std=c++11"]
+if sys.platform == 'darwin':  # Mac os X before Mavericks (10.9)
+  complie_args.append("-stdlib=libc++")
+cppjieba_includes = ["tools/cppjieba/deps",
```
Comment (zh794390558) [inline setup.py]: if add some other tools, how this can work?

File: build_pip_pkg.sh [hunk 13]
```diff
@@ -0,0 +1,8 @@
+#!/usr/bin/env bash
+
+echo "Uninstall tf_jieba if exist ..."
+pip3 uninstall -y delta-didi
```
Comment (zh794390558) [inline build_pip_pkg.sh]: delta

File: build_pip_pkg.sh [hunk 14]
```diff
@@ -0,0 +1,8 @@
...
+
+echo "Generate whl file ..."
+rm -rf build/ delta-didi.egg-info/ dist/
```
Comment (zh794390558) [inline build_pip_pkg.sh]: delta

File: delta/utils/register.py [hunk 15]
```diff
@@ -152,16 +154,31 @@ def _handle_errors(errors):
...
 
 
+def path_to_module_format(py_path):
+  """Transform a python file path to module format."""
+  return py_path.replace("/", ".").rstrip(".py")
+
+
 def import_all_modules_for_register(config=None):
   """Import all modules for register."""
+  current_word_dir = os.getcwd()
```
Comment (zh794390558) [inline delta/utils/register.py]: current_work_dir

File: delta/utils/register.py [hunk 16]
```diff
@@ -152,16 +154,31 @@ def _handle_errors(errors):
     logging.fatal("Module {} import failed: {}".format(name, err))
 
 
+def path_to_module_format(py_path):
```
Comment (zh794390558) [inline delta/utils/register.py]: unit test path_to_module_format

File: delta/utils/register.py [hunk 17]
```diff
@@ -152,16 +154,31 @@ def _handle_errors(errors):
...
+
+
 def import_all_modules_for_register(config=None):
   """Import all modules for register."""
+  current_word_dir = os.getcwd()
+  if current_word_dir not in sys.path:
+    sys.path.append(current_word_dir)
   all_modules = ALL_SYS_MODULES
   errors = []
   if config is not None and "custom_modules" in config:
```
Comment (zh794390558) [inline delta/utils/register.py]: what this for ?

File: delta/utils/register.py [hunk 18]
```diff
@@ -151,7 +146,7 @@ def _handle_errors(errors):
   if not errors:
     return
   for name, err in errors:
-    logging.fatal("Module {} import failed: {}".format(name, err))
```
Comment (zh794390558) [inline delta/utils/register.py]: this should be fatal

File: delta_nlp.egg-info/PKG-INFO [hunk 19]
```diff
@@ -0,0 +1,28 @@
+Metadata-Version: 2.1
```
Comment (zh794390558) [inline delta_nlp.egg-info/PKG-INFO]: remove this file

File: delta/utils/register.py (hunk 17 above)
Comment (applenob) [reply delta/utils/register.py]: This is for customer module import searching.

File: delta/utils/register.py [hunk 20]
```diff
@@ -82,94 +82,159 @@ def __init__(self):
   serving = Register('serving')
 
 
+NLP_TASK_MODULES = [
```
Comment (zh794390558) [inline delta/utils/register.py]: NLP_TASK_MODULES + SPEECH_TASK_MODULES = ALL

File: delta/main.py [hunk 21]
```diff
@@ -105,8 +108,21 @@ def main(argv):
...
-if __name__ == '__main__':
+def entry():
   define_flags()
+  flags.DEFINE_bool('only_nlp', 'False', 'only use nlp modules')
   logging.info("Deep Language Technology Platform start...")
   app.run(main)
   logging.info("OK. Done!")
+
+
+def nlp_entry():
```
Comment (zh794390558) [inline delta/main.py]: same entry and nlp_entry

File: delta/utils/register.py (hunk 20 above)
Comment (applenob) [reply delta/utils/register.py]: Some modules may be both used in nlp and speech.

File: delta/utils/register.py [hunk 22]
```diff
@@ -161,14 +221,38 @@ def _handle_errors(errors):
...
 
-def import_all_modules_for_register():
+def path_to_module_format(py_path):
+  """Transform a python file path to module format."""
+  return py_path.replace("/", ".").rstrip(".py")
+
+
+def import_all_modules_for_register(config=None, only_nlp=False):
   """Import all modules for register."""
+  logging.warning("test")
```
Comment (zh794390558) [inline delta/utils/register.py]: remove

File: delta/main.py (hunk 21 above)
Comment (applenob) [reply delta/main.py]: The `nlp_entry` is for pip `entry_point`, which set `only_nlp = True` by default.

File: delta/utils/register.py (hunk 17 above)
Comment (zh794390558) [reply delta/utils/register.py]: how to using this?

File: delta/layers/ops/Makefile [hunk 23]
```diff
@@ -93,4 +93,4 @@ ${TEST_BIN}: $(TEST_OBJ) $(STATIC_LIB)
 .PHONY: clean
 clean:
 	-rm -rf $(GENDIR)
-	-rm -f x_ops.so
+	-rm -f x_ops.*so
```
Comment (zh794390558) [inline delta/layers/ops/Makefile]: add same case comment

**Pull request merged**